File: src/topic02_motion/turtlebot2/free_space_navigation/free_space_navigation.cpp
/*
 *  Gaitech Educational Portal
 *
 * Copyright (c) 2016
 * All rights reserved.
 *
 * License Type: GNU GPL
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *    Program: Free-Space Navigation
 *
 */

#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "sensor_msgs/LaserScan.h"
#include "nav_msgs/Odometry.h"
#include "tf/tf.h"
#include <tf/transform_listener.h>
#include <fstream>

using namespace std;
#define LINEAR_VELOCITY_MINIMUM_THRESHOLD 0.2
#define ANGULAR_VELOCITY_MINIMUM_THRESHOLD 0.4

//declare publishers
ros::Publisher velocityPublisher;
//declare subscribers
ros::Subscriber scanSubscriber;
ros::Subscriber pose_subscriber;
//global variable to update the position of the robot
nav_msgs::Odometry turtlebot_odom_pose;




//callback function for the /odom topic to update the pose
void poseCallback(const nav_msgs::Odometry::ConstPtr & pose_message);
//the function that makes the robot moves forward and backward
void move_v1(double speed, double distance, bool isForward);
void move_v2(double speed, double distance, bool isForward);
void move_v3(double speed, double distance, bool isForward);
//the function that makes the robot rotates left and right
double rotate(double ang_vel, double angle_radian, bool isClockwise);
double degree2radian(double degreeAngle);
double radian2degree(double radianAngle);
double calculateYaw( double x1, double y1, double x2,double y2);

void moveSquare(double sideLength);

int main(int argc, char **argv){

	//initialize the ROS node
	ros::init(argc, argv, "free_space_navigation_node");
	ros::NodeHandle n;

	//subscribe to the odometry topic to get the position of the robot
	pose_subscriber = n.subscribe("/odom", 10, poseCallback);
	//register the velocity publisher
	velocityPublisher =n.advertise<geometry_msgs::Twist>("/cmd_vel_mux/input/teleop", 1000);

	ros::spinOnce();
	ros::Rate loop(1);
	loop.sleep();loop.sleep();loop.sleep();//loop.sleep();loop.sleep();
	ros::spinOnce();

	while (ros::ok()){
		ros::spinOnce();loop.sleep();
		printf("robot initial pose: (%.2f, %.2f, %.2f)\n",
										turtlebot_odom_pose.pose.pose.position.x,
										turtlebot_odom_pose.pose.pose.position.y,
										radian2degree(tf::getYaw(turtlebot_odom_pose.pose.pose.orientation)));
		moveSquare(1.0);
		//exercise: try to remove the ros::SpinOnce() and observe and comment the result
		ros::spinOnce();loop.sleep();ros::spinOnce();
		printf("robot final pose: (%.2f, %.2f, %.2f)\n",
											turtlebot_odom_pose.pose.pose.position.x,
											turtlebot_odom_pose.pose.pose.position.y,
											radian2degree(tf::getYaw(turtlebot_odom_pose.pose.pose.orientation)));
		return 0;
	}



	return 0;
}




void poseCallback(const nav_msgs::Odometry::ConstPtr & pose_message){
	turtlebot_odom_pose.pose.pose.position.x=pose_message->pose.pose.position.x;
	turtlebot_odom_pose.pose.pose.position.y=pose_message->pose.pose.position.y;
	turtlebot_odom_pose.pose.pose.position.z=pose_message->pose.pose.position.z;

	turtlebot_odom_pose.pose.pose.orientation.w=pose_message->pose.pose.orientation.w;
	turtlebot_odom_pose.pose.pose.orientation.x=pose_message->pose.pose.orientation.x;
	turtlebot_odom_pose.pose.pose.orientation.y=pose_message->pose.pose.orientation.y;
	turtlebot_odom_pose.pose.pose.orientation.z=pose_message->pose.pose.orientation.z;
}

void moveSquare(double sideLength){
	for (int i=0;i<4;i++){
		move_v1(0.3, sideLength, true);
		rotate (0.3, degree2radian(90), true);
	}
}


/**
 * a function that makes the robot move straight
 * @param speed: represents the speed of the robot the robot
 * @param distance: represents the distance to move by the robot
 * @param isForward: if true, the robot moves forward,otherwise, it moves backward
 *
 * Method 1: using tf and Calculate the distance between the two transformations
 */
void move_v1(double speed, double distance, bool isForward){
	//declare a Twist message to send velocity commands
	geometry_msgs::Twist VelocityMessage;
	//declare tf transform listener: this transform listener will be used to listen and capture the transformation between
	// the /odom frame (that represent the reference frame) and the base_footprint frame the represent moving frame
	tf::TransformListener listener;
	//declare tf transform
	//init_transform: is the transformation before starting the motion
	tf::StampedTransform init_transform;
	//current_transformation: is the transformation while the robot is moving
	tf::StampedTransform current_transform;


	//set the linear velocity to a positive value if isFoward is true
	if (isForward)
		VelocityMessage.linear.x =abs(speed);
	else //else set the velocity to negative value to move backward
		VelocityMessage.linear.x =-abs(speed);
	//all velocities of other axes must be zero.
	VelocityMessage.linear.y =0;
	VelocityMessage.linear.z =0;
	//The angular velocity of all axes must be zero because we want  a straight motion
	VelocityMessage.angular.x = 0;
	VelocityMessage.angular.y = 0;
	VelocityMessage.angular.z =0;

	double distance_moved = 0.0;
	ros::Rate loop_rate(10); // we publish the velocity at 10 Hz (10 times a second)

	/*
	 * First, we capture the initial transformation before starting the motion.
	 * we call this transformation "init_transform"
	 * It is important to "waitForTransform" otherwise, it might not be captured.
	 */
	try{
		//wait for the transform to be found
		listener.waitForTransform("/base_footprint", "/odom", ros::Time(0), ros::Duration(10.0) );
		//Once the transform is found,get the initial_transform transformation.
		listener.lookupTransform("/base_footprint", "/odom",ros::Time(0), init_transform);
	}
	catch (tf::TransformException & ex){
		ROS_ERROR(" Problem %s",ex.what());
		ros::Duration(1.0).sleep();
	}



	do{
		/***************************************
		 * STEP1. PUBLISH THE VELOCITY MESSAGE
		 ***************************************/
		velocityPublisher.publish(VelocityMessage);
		ros::spinOnce();
		loop_rate.sleep();
		/**************************************************
		 * STEP2. ESTIMATE THE DISTANCE MOVED BY THE ROBOT
		 *************************************************/
		try{
			//wait for the transform to be found
			listener.waitForTransform("/base_footprint", "/odom", ros::Time(0), ros::Duration(10.0) );
			//Once the transform is found,get the initial_transform transformation.
			listener.lookupTransform("/base_footprint", "/odom",ros::Time(0), current_transform);
		}
		catch (tf::TransformException & ex){
			ROS_ERROR(" Problem %s",ex.what());
			ros::Duration(1.0).sleep();
		}
		/*
		 * Calculate the distance moved by the robot
		 * There are two methods that give the same result
		 */

		/*
		 * Method 1: Calculate the distance between the two transformations
		 * Hint:
		 * 	  --> transform.getOrigin().x(): represents the x coordinate of the transformation
		 * 	  --> transform.getOrigin().y(): represents the y coordinate of the transformation
		 */
		//calculate the distance moved
		//cout<<"Initial Transform: "<<init_transform <<" , "<<"Current Transform: "<<current_transform<<endl;

		distance_moved = sqrt(pow((current_transform.getOrigin().x()-init_transform.getOrigin().x()), 2) +
				pow((current_transform.getOrigin().y()-init_transform.getOrigin().y()), 2));


	}while((distance_moved<distance)&&(ros::ok()));
	//finally, stop the robot when the distance is moved
	VelocityMessage.linear.x =0;
	velocityPublisher.publish(VelocityMessage);
}

/**
 * a function that makes the robot move straight
 * @param speed: represents the speed of the robot the robot
 * @param distance: represents the distance to move by the robot
 * @param isForward: if true, the robot moves forward,otherwise, it moves backward
 *
 * Method 2: using tf and we calculate the relative transform, then we determine its length
 */
void move_v2(double speed, double distance, bool isForward){
	//delcare a Twist message to send velocity commands
	geometry_msgs::Twist VelocityMessage;
	//declare tf transform listener: this transform listener will be used to listen and capture the transformation between
	// the odom frame (that represent the reference frame) and the base_footprint frame the represent moving frame
	tf::TransformListener listener;
	//declare tf transform
	//init_transform: is the transformation before starting the motion
	tf::StampedTransform init_transform;
	//current_transformation: is the transformation while the robot is moving
	tf::StampedTransform current_transform;
	//initial coordinates (for method 3)
	nav_msgs::Odometry initial_turtlebot_odom_pose;

	//set the linear velocity to a positive value if isFoward is true
	if (isForward)
		VelocityMessage.linear.x =abs(speed);
	else //else set the velocity to negative value to move backward
		VelocityMessage.linear.x =-abs(speed);
	//all velocities of other axes must be zero.
	VelocityMessage.linear.y = VelocityMessage.linear.z =VelocityMessage.angular.x =VelocityMessage.angular.y =VelocityMessage.angular.z =0;

	double distance_moved = 0.0;
	ros::Rate loop_rate(10); // we publish the velocity at 10 Hz (10 times a second)

	/*
	 * First, we capture the initial transformation before starting the motion.
	 * we call this transformation "init_transform"
	 * It is important to "waitForTransform" otherwise, it might not be captured.
	 */
	try{
		//wait for the transform to be found
		listener.waitForTransform("/base_footprint", "/odom", ros::Time(0), ros::Duration(10.0) );
		//Once the transform is found,get the initial_transform transformation.
		listener.lookupTransform("/base_footprint", "/odom",ros::Time(0), init_transform);
	}
	catch (tf::TransformException & ex){
		ROS_ERROR(" Problem %s",ex.what());
		ros::Duration(1.0).sleep();
	}



	do{
		/***************************************
		 * STEP1. PUBLISH THE VELOCITY MESSAGE
		 ***************************************/
		velocityPublisher.publish(VelocityMessage);
		ros::spinOnce();
		loop_rate.sleep();
		/**************************************************
		 * STEP2. ESTIMATE THE DISTANCE MOVED BY THE ROBOT
		 *************************************************/
		try{
			//wait for the transform to be found
			listener.waitForTransform("/base_footprint", "/odom", ros::Time(0), ros::Duration(10.0) );
			//Once the transform is found,get the initial_transform transformation.
			listener.lookupTransform("/base_footprint", "/odom",ros::Time(0), current_transform);
		}
		catch (tf::TransformException & ex){
			ROS_ERROR(" Problem %s",ex.what());
			ros::Duration(1.0).sleep();
		}

		/*
		 * Method 2: using transform composition. We calculate the relative transform, then we determine its length
		 * Hint:
		 * 	  --> transform.getOrigin().length(): return the displacement of the origin of the transformation
		 */
		tf::Transform relative_transform = init_transform.inverse() * current_transform;
		distance_moved= relative_transform.getOrigin().length();

		//cout<<"Method 2: distance moved: "<<distance_moved <<", "<<distance<<endl;

	}while((distance_moved<distance)&&(ros::ok()));
	//finally, stop the robot when the distance is moved
	VelocityMessage.linear.x =0;
	velocityPublisher.publish(VelocityMessage);
}


/**
 * a function that makes the robot move straight
 * @param speed: represents the speed of the robot the robot
 * @param distance: represents the distance to move by the robot
 * @param isForward: if true, the robot moves forward,otherwise, it moves backward
 *
 * Method 3: we use coordinates of the robot to estimate the distance
 */
void move_v3(double speed, double distance, bool isForward){
	//delcare a Twist message to send velocity commands
	geometry_msgs::Twist VelocityMessage;
	//initial pose of the turtlebot before start moving
	nav_msgs::Odometry initial_turtlebot_odom_pose;

	//set the linear velocity to a positive value if isFoward is true
	if (isForward)
		VelocityMessage.linear.x =abs(speed);
	else //else set the velocity to negative value to move backward
		VelocityMessage.linear.x =-abs(speed);
	//all velocities of other axes must be zero.
	VelocityMessage.linear.y = VelocityMessage.linear.z =VelocityMessage.angular.x =VelocityMessage.angular.y =VelocityMessage.angular.z =0;

	double distance_moved = 0.0;
	ros::Rate loop_rate(10); // we publish the velocity at 10 Hz (10 times a second)

	//we update the initial_turtlebot_odom_pose using the turtlebot_odom_pose global variable updated in the callback function poseCallback
	initial_turtlebot_odom_pose = turtlebot_odom_pose;

	do{
		velocityPublisher.publish(VelocityMessage);
		ros::spinOnce();
		loop_rate.sleep();
		distance_moved = sqrt(pow((turtlebot_odom_pose.pose.pose.position.x-initial_turtlebot_odom_pose.pose.pose.position.x), 2) +
				pow((turtlebot_odom_pose.pose.pose.position.y-initial_turtlebot_odom_pose.pose.pose.position.y), 2));

	}while((distance_moved<distance)&&(ros::ok()));
	//finally, stop the robot when the distance is moved
	VelocityMessage.linear.x =0;
	velocityPublisher.publish(VelocityMessage);
}



double rotate(double angular_velocity, double radians,  bool clockwise)
{

	//delcare a Twist message to send velocity commands
	geometry_msgs::Twist VelocityMessage;
	//declare tf transform listener: this transform listener will be used to listen and capture the transformation between
	// the odom frame (that represent the reference frame) and the base_footprint frame the represent moving frame
	tf::TransformListener TFListener;
	//declare tf transform
	//init_transform: is the transformation before starting the motion
	tf::StampedTransform init_transform;
	//current_transformation: is the transformation while the robot is moving
	tf::StampedTransform current_transform;
	//initial coordinates (for method 3)
	nav_msgs::Odometry initial_turtlebot_odom_pose;

	double angle_turned =0.0;

	//validate angular velocity; ANGULAR_VELOCITY_MINIMUM_THRESHOLD is the minimum allowed
	angular_velocity=((angular_velocity>ANGULAR_VELOCITY_MINIMUM_THRESHOLD)?angular_velocity:ANGULAR_VELOCITY_MINIMUM_THRESHOLD);

	while(radians < 0) radians += 2*M_PI;
	while(radians > 2*M_PI) radians -= 2*M_PI;

	//wait for the listener to get the first message
	TFListener.waitForTransform("base_footprint", "odom", ros::Time(0), ros::Duration(1.0));


	//record the starting transform from the odometry to the base frame
	TFListener.lookupTransform("base_footprint", "odom", ros::Time(0), init_transform);


	//the command will be to turn at 0.75 rad/s
	VelocityMessage.linear.x = VelocityMessage.linear.y = 0.0;
	VelocityMessage.angular.z = angular_velocity;
	if (clockwise) VelocityMessage.angular.z = -VelocityMessage.angular.z;

	//the axis we want to be rotating by
	tf::Vector3 desired_turn_axis(0,0,1);
	if (!clockwise) desired_turn_axis = -desired_turn_axis;

	ros::Rate rate(10.0);
	bool done = false;
	while (!done )
	{
		//send the drive command
		velocityPublisher.publish(VelocityMessage);
		rate.sleep();
		//get the current transform
		try
		{
			TFListener.waitForTransform("base_footprint", "odom", ros::Time(0), ros::Duration(1.0));
			TFListener.lookupTransform("base_footprint", "odom", ros::Time(0), current_transform);
		}
		catch (tf::TransformException ex)
		{
			ROS_ERROR("%s",ex.what());
			break;
		}
		tf::Transform relative_transform = init_transform.inverse() * current_transform;
		tf::Vector3 actual_turn_axis = relative_transform.getRotation().getAxis();
		angle_turned = relative_transform.getRotation().getAngle();

		if (fabs(angle_turned) < 1.0e-2) continue;
		if (actual_turn_axis.dot(desired_turn_axis ) < 0 )
			angle_turned = 2 * M_PI - angle_turned;

		if (!clockwise)
			VelocityMessage.angular.z = (angular_velocity-ANGULAR_VELOCITY_MINIMUM_THRESHOLD) * (fabs(radian2degree(radians-angle_turned)/radian2degree(radians)))+ANGULAR_VELOCITY_MINIMUM_THRESHOLD;
		else
			if (clockwise)
				VelocityMessage.angular.z = (-angular_velocity+ANGULAR_VELOCITY_MINIMUM_THRESHOLD) * (fabs(radian2degree(radians-angle_turned)/radian2degree(radians)))-ANGULAR_VELOCITY_MINIMUM_THRESHOLD;

		if (angle_turned > radians) {
			done = true;
			VelocityMessage.linear.x = VelocityMessage.linear.y = VelocityMessage.angular.z = 0;
			velocityPublisher.publish(VelocityMessage);
		}


	}
	if (done) return angle_turned;
	return angle_turned;
}


double calculateYaw( double x1, double y1, double x2,double y2)
{

	double bearing = atan2((y2 - y1),(x2 - x1));
	//if(bearing < 0) bearing += 2 * PI;
	bearing *= 180.0 / M_PI;
	return bearing;
}

/* makes conversion from radian to degree */
double radian2degree(double radianAngle){
	return (radianAngle*57.2957795);
}


/* makes conversion from degree to radian */
double degree2radian(double degreeAngle){
	return (degreeAngle/57.2957795);
}

File: src/topic02_motion/turtlebot2/free_space_navigation/free_space_navigation.py
#!/usr/bin/env python

import rospy
import tf
import numpy
import geometry_msgs.msg
from geometry_msgs.msg import Twist
from math import *
from nav_msgs.msg import Odometry
from std_msgs.msg import String
import tf2_ros
import copy

LINEAR_VELOCITY_MINIMUM_THRESHOLD  = 0.2
ANGULAR_VELOCITY_MINIMUM_THRESHOLD = 0.4
class free_space_navigation():

    

    def poseCallback(self,pose_message):

        self.turtlebot_odom_pose.pose.pose.position.x=pose_message.pose.pose.position.x
        self.turtlebot_odom_pose.pose.pose.position.y=pose_message.pose.pose.position.y
        self.turtlebot_odom_pose.pose.pose.position.z=pose_message.pose.pose.position.z

        self.turtlebot_odom_pose.pose.pose.orientation.w=pose_message.pose.pose.orientation.w
        self.turtlebot_odom_pose.pose.pose.orientation.x=pose_message.pose.pose.orientation.x
        self.turtlebot_odom_pose.pose.pose.orientation.y=pose_message.pose.pose.orientation.y
        self.turtlebot_odom_pose.pose.pose.orientation.z=pose_message.pose.pose.orientation.z

 # a function that makes the robot move straight
 # @param speed: represents the speed of the robot the robot
 # @param distance: represents the distance to move by the robot
 # @param isForward: if True, the robot moves forward,otherwise, it moves backward
 #
 # Method 1: using tf and Calculate the distance between the two transformations

    def move_v1(self, speed, distance, isForward):
        #declare a Twist message to send velocity commands
        VelocityMessage = Twist()
        # declare tf transform listener: this transform listener will be used to listen and capture the transformation between
        # the odom frame (that represent the reference frame) and the base_footprint frame the represent moving frame
        listener = tf.TransformListener()

        #set the linear velocity to a positive value if isFoward is True
        if (isForward):
            VelocityMessage.linear.x =abs(speed)
        else: #else set the velocity to negative value to move backward
            VelocityMessage.linear.x =-abs(speed)

        # all velocities of other axes must be zero.
        VelocityMessage.linear.y =0
        VelocityMessage.linear.z =0
        #The angular velocity of all axes must be zero because we want  a straight motion
        VelocityMessage.angular.x = 0
        VelocityMessage.angular.y = 0
        VelocityMessage.angular.z =0

        distance_moved = 0.0
        loop_rate = rospy.Rate(10) # we publish the velocity at 10 Hz (10 times a second)    


     #  First, we capture the initial transformation before starting the motion.
     # we call this transformation "init_transform"
     # It is important to "waitForTransform" otherwise, it might not be captured.

        try:
            #wait for the transform to be found

            listener.waitForTransform("/base_footprint", "/odom", rospy.Time(0),rospy.Duration(10.0))
            #Once the transform is found,get the initial_transform transformation.
            (trans,rot) = listener.lookupTransform('/base_footprint', '/odom', rospy.Time(0))
            #listener.lookupTransform("/base_footprint", "/odom", rospy.Time(0),init_transform)
            start = 0.5 * sqrt(trans[0] ** 2 + trans[1] ** 2)

        except Exception:
            rospy.Duration(1.0)

        distance_moved = 0
        
        while True :
            rospy.loginfo("Turtlebot moves forwards") 
        #/***************************************
        # * STEP1. PUBLISH THE VELOCITY MESSAGE
        # ***************************************/
            self.velocityPublisher.publish(VelocityMessage)
            loop_rate.sleep()
        #/**************************************************
        # * STEP2. ESTIMATE THE DISTANCE MOVED BY THE ROBOT
        # *************************************************/
            try:

                #wait for the transform to be found
                listener.waitForTransform("/base_footprint", "/odom", rospy.Time(0), rospy.Duration(10.0) )
                #Once the transform is found,get the initial_transform transformation.
                #listener.lookupTransform("/base_footprint", "/odom",rospy.Time(0))
                (trans,rot) = listener.lookupTransform('/base_footprint', '/odom', rospy.Time(0))
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
                rospy.Duration(1.0)
        # Calculate the distance moved by the robot
        # There are two methods that give the same result
        #
        # Method 1: Calculate the distance between the two transformations
        # Hint:
        #    --> transform.getOrigin().x(): represents the x coordinate of the transformation
        #    --> transform.getOrigin().y(): represents the y coordinate of the transformation
        #
        # calculate the distance moved
            end = 0.5 * sqrt(trans[0] ** 2 + trans[1] ** 2)
            distance_moved = distance_moved+abs(abs(float(end)) - abs(float(start)))
            if not (distance_moved<distance):
                break
            
            #finally, stop the robot when the distance is moved
        VelocityMessage.linear.x =0 
        self.velocityPublisher.publish(VelocityMessage)

    
    def move_v2(self, speed, distance, isForward):

        #declare a Twist message to send velocity commands
        VelocityMessage = Twist()
        # declare tf transform listener: this transform listener will be used to listen and capture the transformation between
        # the odom frame (that represent the reference frame) and the base_footprint frame the represent moving frame
        # set the linear velocity to a positive value if isFoward is True
        listener = tf.TransformListener()
        #declare tf transform
        initial_turtlebot_odom_pose = Odometry()
        #init_transform: is the transformation before starting the motion
        init_transform = geometry_msgs.msg.TransformStamped()
        #current_transformation: is the transformation while the robot is moving
        current_transform = geometry_msgs.msg.TransformStamped()

        if (isForward):
            VelocityMessage.linear.x =abs(speed)
        else: #else set the velocity to negative value to move backward
            VelocityMessage.linear.x =-abs(speed)
        #all velocities of other axes must be zero.
        VelocityMessage.linear.y =0.0
        VelocityMessage.linear.z =0.0
        VelocityMessage.angular.x =0.0
        VelocityMessage.angular.y =0.0
        VelocityMessage.angular.z =0.0

        distance_moved = 0.0

        loop_rate = rospy.Rate(10) # we publish the velocity at 10 Hz (10 times a second)

     # First, we capture the initial transformation before starting the motion.
     # we call this transformation "init_transform"
     # It is important to "waitForTransform" otherwise, it might not be captured.
        try:
            #wait for the transform to be found

            listener.waitForTransform("/base_footprint", "/odom", rospy.Time(0),rospy.Duration(10.0))
            #Once the transform is found,get the initial_transform transformation.
            (trans,rot) = listener.lookupTransform('/base_footprint', '/odom', rospy.Time(0))
            #listener.lookupTransform("/base_footprint", "/odom", rospy.Time(0),init_transform)
            trans1_mat = tf.transformations.translation_matrix(trans)
            rot1_mat   = tf.transformations.quaternion_matrix(rot)
            mat1 = numpy.dot(trans1_mat, rot1_mat)
            init_transform.transform.translation = trans
            init_transform.transform.rotation =rot
            #print(mat1)
        except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
            rospy.Duration(1.0)
     
        while True :
            rospy.loginfo("Turtlebot moves forwards") 
        #/***************************************
        # * STEP1. PUBLISH THE VELOCITY MESSAGE
        # ***************************************/
            self.velocityPublisher.publish(VelocityMessage)
            loop_rate.sleep()
        #/**************************************************
        # * STEP2. ESTIMATE THE DISTANCE MOVED BY THE ROBOT
        # *************************************************/
            try:

                #wait for the transform to be found
                listener.waitForTransform("/base_footprint", "/odom", rospy.Time(0), rospy.Duration(10.0) )
                #Once the transform is found,get the initial_transform transformation.
                #listener.lookupTransform("/base_footprint", "/odom",rospy.Time(0))
                (trans,rot) = listener.lookupTransform('/base_footprint', '/odom', rospy.Time(0))
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
                rospy.Duration(1.0)
            
            trans1_mat = tf.transformations.translation_matrix(trans)
            rot1_mat   = tf.transformations.quaternion_matrix(rot)
            mat2 = numpy.dot(trans1_mat, rot1_mat)
            #print(mat2)
            mat3 = numpy.dot(mat1, mat2)
            #print(mat3)

            trans3 = tf.transformations.translation_from_matrix(mat3)
            
            #print(trans3)
            rot3 = tf.transformations.quaternion_from_matrix(mat3)
            #print(rot3)
            
            current_transform.transform.translation = trans
            current_transform.transform.rotation =rot
            distance_moved = distance_moved + (0.5 * sqrt(trans3[0] ** 2 + trans3[1] ** 2))
            #print(distance_moved)
            #print (numpy.linalg.norm(trans3))
            
            if not (distance_moved<distance):
                break
            
    #finally, stop the robot when the distance is moved
        VelocityMessage.linear.x =0
        self.velocityPublisher.publish(VelocityMessage)
    


    # a function that makes the robot move straight
    # @param speed: represents the speed of the robot the robot
    # @param distance: represents the distance to move by the robot
    # @param isForward: if True, the robot moves forward,otherwise, it moves backward
    #
    # Method 3: we use coordinates of the robot to estimate the distance

    def move_v3(self, speed, distance, isForward):
        #declare a Twist message to send velocity commands
            VelocityMessage = Twist()
        # declare tf transform listener: this transform listener will be used to listen and capture the transformation between
        # the odom frame (that represent the reference frame) and the base_footprint frame the represent moving frame
            listener = tf.TransformListener()
        #declare tf transform
            initial_turtlebot_odom_pose = Odometry()
        #init_transform: is the transformation before starting the motion
            init_transform = geometry_msgs.msg.TransformStamped()
        #current_transformation: is the transformation while the robot is moving
            current_transform = geometry_msgs.msg.TransformStamped()

        #set the linear velocity to a positive value if isFoward is True
            if (isForward):
                    VelocityMessage.linear.x =abs(speed)
            else: #else set the velocity to negative value to move backward
                    VelocityMessage.linear.x =-abs(speed)

        # all velocities of other axes must be zero.
            VelocityMessage.linear.y =0
            VelocityMessage.linear.z =0
        #The angular velocity of all axes must be zero because we want  a straight motion
            VelocityMessage.angular.x = 0
            VelocityMessage.angular.y = 0
            VelocityMessage.angular.z = 0

            distance_moved = 0.0
            loop_rate = rospy.Rate(20) # we publish the velocity at 10 Hz (10 times a second)    
            #we update the initial_turtlebot_odom_pose using the turtlebot_odom_pose global variable updated in the callback function poseCallback
            #we will use deepcopy() to avoid pointers confusion
            initial_turtlebot_odom_pose = copy.deepcopy(self.turtlebot_odom_pose)

            while True :
                    rospy.loginfo("Turtlebot moves forwards")
                    self.velocityPublisher.publish(VelocityMessage)
         
                    loop_rate.sleep()
                    
                    #rospy.Duration(1.0)
                    
                    distance_moved = distance_moved+abs(0.5 * sqrt(((self.turtlebot_odom_pose.pose.pose.position.x-initial_turtlebot_odom_pose.pose.pose.position.x) ** 2) +
                        ((self.turtlebot_odom_pose.pose.pose.position.x-initial_turtlebot_odom_pose.pose.pose.position.x) ** 2)))
                                        
                    if not (distance_moved<distance):
                        break
            
            #finally, stop the robot when the distance is moved
            VelocityMessage.linear.x =0
            self.velocityPublisher.publish(VelocityMessage)

    def degree2radian(self, degreeAngle):
        return (degreeAngle/57.2957795)
    
    def rotate(self,angular_velocity,radians,clockwise):
        rotateMessage = Twist()
       
        #declare tf transform 
        listener = tf.TransformListener()
        #init_transform: is the transformation before starting the motion
        init_transform = geometry_msgs.msg.TransformStamped()
        #current_transformation: is the transformation while the robot is moving
        current_transform = geometry_msgs.msg.TransformStamped()
        
        angle_turned = 0.0

        angular_velocity = (-angular_velocity, ANGULAR_VELOCITY_MINIMUM_THRESHOLD)[angular_velocity > ANGULAR_VELOCITY_MINIMUM_THRESHOLD]

        while(radians < 0):
            radians += 2* pi

        while(radians > 2* pi):
            radians -= 2* pi
        
        listener.waitForTransform("/base_footprint", "/odom", rospy.Time(0), rospy.Duration(10.0) )
        (trans,rot) = listener.lookupTransform('/base_footprint', '/odom', rospy.Time(0))
        #listener.lookupTransform("/base_footprint", "/odom", rospy.Time(0),init_transform)
        
        init_transform.transform.translation = trans
        init_transform.transform.rotation =rot

        #since the rotation is only in the Z-axes 
        #start_angle = tf.transformations.#0.5 * sqrt(rot[2] ** 2)
        euler = tf.transformations.euler_from_quaternion(rot)
        roll = euler[0]
        pitch = euler[1]
        start_angle = euler[2]

        rotateMessage.linear.x = rotateMessage.linear.y = 0.0
        rotateMessage.angular.z = angular_velocity

        if (clockwise):
            rotateMessage.angular.z = -rotateMessage.angular.z
        
        
        loop_rate = rospy.Rate(20)
        
        while True:
            rospy.loginfo("Turtlebot is Rotating")

            self.velocityPublisher.publish(rotateMessage)
         
            loop_rate.sleep()
                    
            #rospy.Duration(1.0)

            try:

                #wait for the transform to be found
                listener.waitForTransform("/base_footprint", "/odom", rospy.Time(0), rospy.Duration(10.0) )
                #Once the transform is found,get the initial_transform transformation.
                #listener.lookupTransform("/base_footprint", "/odom",rospy.Time(0))
                (trans,rot) = listener.lookupTransform('/base_footprint', '/odom', rospy.Time(0))
            except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
                rospy.Duration(1.0)

            current_transform.transform.translation = trans
            current_transform.transform.rotation =rot

            #since the rotation is only in the Z-axes 
            #end_angle = 0.5 * sqrt( rot[2] ** 2)
            euler = tf.transformations.euler_from_quaternion(rot)
            roll = euler[0]
            pitch = euler[1]
            end_angle = euler[2]
            
            angle_turned = abs(end_angle - start_angle)
            print "angle_turned: %s" %angle_turned
            if (angle_turned > radians):
                break

        self.velocityPublisher.publish(rotateMessage)
        loop_rate.sleep()

    def calculateYaw(self, x1, y1, x2,y2):
        bearing = atan2((y2 - y1),(x2 - x1))
        #if(bearing < 0) bearing += 2 * PI
        bearing *= 180.0 / pi
        return bearing


    def moveSquare(self,sideLength):
        for i in range(0, 4):
            self.move_v1(0.3, sideLength, True)
            #self.shutdown()
            self.rotate(0.2, self.degree2radian(90.0),True)
   
    def __init__(self):
        # initiliaze
        rospy.init_node('free_space_navigation', anonymous=True)

        # What to do you ctrl + c    
        rospy.on_shutdown(self.shutdown)
        self.turtlebot_odom_pose = Odometry()
        pose_message = Odometry()
        self.velocityPublisher = rospy.Publisher('/cmd_vel_mux/input/teleop', Twist, queue_size=10)
        self.pose_subscriber = rospy.Subscriber("/odom", Odometry, self.poseCallback)     
        # 2 HZ
        r = rospy.Rate(20)
        
        r.sleep()

        while not rospy.is_shutdown():
            sideLength=1
            self.moveSquare(sideLength)
            #self.rotate(0.3, self.degree2radian(90.0), True);
            #self.rotate(0.3, self.degree2radian(90.0), False);       

        
    def shutdown(self):
        # stop turtlebot
        rospy.loginfo("Stop Drawing Squares")
        self.velocityPublisher.publish(Twist())
        rospy.sleep(1)
    
 

    

if __name__ == '__main__':
    try:
        free_space_navigation()
        rospy.spin()
    except rospy.ROSInterruptException:
        rospy.loginfo("node terminated.")
